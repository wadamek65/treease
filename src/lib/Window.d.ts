declare global {
	interface Window {
		IS_STORYBOOK: boolean;
	}
}

export {};
