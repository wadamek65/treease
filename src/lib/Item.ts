export type ItemTypes = 'file' | 'directory';

export type Item = {
	id: string;
	name: string;
	level: number;
	itemType: ItemTypes;
	isCollapsed: boolean;
	parentId: string;
};
