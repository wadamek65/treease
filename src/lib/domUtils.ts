export function getActiveItemElement(): Element | null {
	return document.querySelector('[data-tree-item]:focus');
}

export function getAllItemElements(): HTMLElement[] {
	return Array.from(document.querySelectorAll('[data-tree-item]'));
}

export function getParentItemElement(): HTMLElement | null | undefined {
	const activeItemElement = getActiveItemElement();
	if (!activeItemElement) {
		return null;
	}

	return activeItemElement.parentNode?.querySelector('[data-tree-item]');
}

export function getActiveItemElementIndex(): number {
	const activeItemElement = getActiveItemElement();
	if (!activeItemElement) {
		return -1;
	}

	const allItemElements = getAllItemElements();
	return allItemElements.findIndex((treeItem) => treeItem.isEqualNode(activeItemElement));
}

export function focusClosestActiveElement(fallback: 'last' | 'first'): void {
	const closestActiveItemElement = document.activeElement?.closest<HTMLElement>('[data-tree-item]');
	if (closestActiveItemElement) {
		closestActiveItemElement.focus();
	} else {
		const allItemElements = getAllItemElements();
		if (allItemElements.length > 0) {
			allItemElements[fallback === 'first' ? 0 : allItemElements.length - 1].focus();
		}
	}
}

export function focusFirstChildItemElement(): void {
	const activeElement = getActiveItemElement();
	if (!activeElement) {
		return;
	}

	const activeElementLevel = activeElement.attributes.getNamedItem('data-tree-item-level');
	if (!activeElementLevel) {
		return;
	}

	const nextLevel = Number(activeElementLevel.value) + 1;
	activeElement.parentNode
		?.querySelector<HTMLElement>(`[data-tree-item-level="${nextLevel}"]`)
		?.focus();
}

export function focusItemElement(id: string): void {
	document.querySelector<HTMLElement>(`[data-tree-item="${id}"]`)?.focus();
}
