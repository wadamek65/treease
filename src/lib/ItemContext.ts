import { createContext } from 'solid-js';
import { Item } from '~/lib/Item';

export const ItemContext = createContext<Item>();
