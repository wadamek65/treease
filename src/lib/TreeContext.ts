import { createContext } from 'solid-js';
import { ItemTypes, Item } from '~/lib/Item';

export type TreeStore = {
	items: Record<string, Item>;
	isPrinting: boolean;
};

type TreeContextValue = {
	treeStore: TreeStore;
	createNewItem: (parentId: string, itemType: ItemTypes) => void;
	removeItem: (id: string) => void;
	renameItem: (id: string, newName: string) => void;
	collapseItem: (id: string) => void;
	expandItem: (id: string) => void;
	toggleCollapsed: (id: string) => void;
	changeParent: (id: string, newParentId: string) => void;
	duplicateItem: (id: string) => void;
	setIsPrinting: (updater: ((isPrinting: boolean) => boolean) | boolean) => void;
};

export const TreeContext = createContext<TreeContextValue>();
