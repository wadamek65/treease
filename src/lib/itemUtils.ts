import { Item } from '~/lib/Item';

export const rootId = 'root';

export function isRoot(item: Item) {
	return item.id === rootId;
}

export function isRootId(itemId: string) {
	return itemId === rootId;
}

export function isFile(item: Item) {
	return item.itemType === 'file';
}

export function isDirectory(item: Item) {
	return item.itemType === 'directory';
}
