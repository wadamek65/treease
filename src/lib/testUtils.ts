import { Item } from '~/lib/Item';
import uniqid from 'uniqid';

export function file(name: string, isCollapsed: boolean = false, parentId?: string): Partial<Item> {
	return {
		id: uniqid.time(),
		name,
		itemType: 'file',
		isCollapsed,
		parentId,
		level: 1,
	};
}

export function directory(
	name: string,
	isCollapsed: boolean = false,
	parentId?: string,
	customId?: string,
): {
	children: (
		items: (ReturnType<typeof directory> | ReturnType<typeof file>)[],
	) => ReturnType<typeof directory>;
	toItems: () => Record<string, Item>;
	items: Partial<Item>[];
} {
	const id = customId ?? uniqid.time();

	const thisItem: Partial<Item> = {
		id,
		name,
		itemType: 'directory',
		isCollapsed,
		parentId,
		level: 1,
	};

	const itemsToReturn = [thisItem];

	return {
		items: itemsToReturn,
		toItems() {
			return this.items.reduce(
				(acc, value) => ({ ...acc, [value.id as string]: value as Item }),
				{} as Record<string, Item>,
			);
		},
		children(items: (ReturnType<typeof directory> | ReturnType<typeof file>)[]) {
			items.forEach((item) => {
				if ('items' in item) {
					item.items[0].parentId = id;
					item.items.forEach((item) => ((item.level as number) += 1));
					itemsToReturn.push(...item.items);
				} else {
					item.parentId = id;
					(item.level as number) += 1;
					itemsToReturn.push(item);
				}
			});
			return this;
		},
	};
}
