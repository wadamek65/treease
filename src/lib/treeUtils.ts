import { TreeStore } from '~/lib/TreeContext';
import { Item } from '~/lib/Item';
import { rootId } from './itemUtils';

type PartialItem = Pick<Item, 'name' | 'itemType' | 'level'>;
export type ItemWithChildren = PartialItem & { children: ItemWithChildren[] };
export function treeToJSON(tree: TreeStore): ItemWithChildren {
	function toJSONRecursive(itemId: string): ItemWithChildren {
		const item = tree.items[itemId];
		return {
			name: item.name,
			itemType: item.itemType,
			level: item.level,
			children: Object.values(tree.items)
				.filter((child) => child.parentId === itemId)
				.map((item) => toJSONRecursive(item.id)),
		};
	}

	return toJSONRecursive(rootId);
}

export function childrenToSortedByTypeName(tree: TreeStore, parentId: string): string[] {
	return Object.values(tree.items)
		.filter((item) => item.parentId === parentId)
		.toSorted((a, b) => a.name.localeCompare(b.name))
		.toSorted((a, b) => (a.itemType === 'directory' ? (b.itemType === 'file' ? -1 : 0) : 1))
		.map((item) => item.id);
}

export function itemsToSortedByTypeName(items: ItemWithChildren[]): ItemWithChildren[] {
	return items
		.toSorted((a, b) => a.name.localeCompare(b.name))
		.toSorted((a, b) => (a.itemType === 'directory' ? (b.itemType === 'file' ? -1 : 0) : 1));
}

export function treeToSortedByLevelTypeName(tree: TreeStore): Item[] {
	return Object.values(tree.items)
		.toSorted((a, b) => a.name.localeCompare(b.name))
		.toSorted((a, b) => (a.level === b.level ? 0 : a.level > b.level ? 1 : -1))
		.toSorted((a, b) => (a.itemType === 'directory' ? (b.itemType === 'file' ? -1 : 0) : 1))
		.map((item) => item);
}
