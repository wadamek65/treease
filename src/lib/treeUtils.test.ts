import { expect, test } from 'vitest';
import { treeToJSON, JSONToASCII } from './treeUtils';
import { TreeStore } from '~/lib/TreeContext';
import { rootId } from '~/lib/itemUtils';
import { directory, file } from '~/lib/testUtils';

const treeToTest1: TreeStore = {
	isPrinting: false,
	items: directory('/root', false, '', rootId)
		.children([
			directory('a_dir', false, rootId).children([
				file('a_file.txt'),
				directory('a_dir').children([file('file.txt')]),
				file('c_file.txt'),
				directory('b_dir').children([file('file.txt')]),
				file('b_file.txt'),
			]),
		])
		.toItems(),
};

const treeToTest2: TreeStore = {
	isPrinting: false,
	items: directory('/root', false, '', rootId)
		.children([
			directory('a_dir', true, rootId).children([
				file('a_file.txt'),
				file('c_file.txt'),
				file('b_file.txt'),
			]),
		])
		.toItems(),
};

const treeToTest3: TreeStore = {
	isPrinting: false,
	items: directory('/root', false, '', rootId)
		.children([
			directory('dir1', false, rootId).children([
				file('file1.txt'),
				file('file2.txt'),
				directory('dir2').children([file('file3.txt')]),
			]),
		])
		.toItems(),
};

const tests = [treeToTest1, treeToTest2, treeToTest3];

test.each(tests)('treeToJSON tests - %#', (treeToTest) => {
	const result = treeToJSON(treeToTest);
	// treeToJSON does not sort the children
	expect(result).toMatchSnapshot();
});

test.each(tests)('JSONToASCII tests - %#', (treeToTest) => {
	const result = JSONToASCII(treeToJSON(treeToTest));
	// treeToJSON does not sort the children
	expect(result).toMatchSnapshot();
});
