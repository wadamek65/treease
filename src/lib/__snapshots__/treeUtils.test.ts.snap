// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSONToASCII tests - 0 1`] = `
"┌ /root
├── a_dir/
├─── a_dir/
├──── file.txt
├─── b_dir/
├──── file.txt
├─── a_file.txt
├─── b_file.txt
├─── c_file.txt
"
`;

exports[`JSONToASCII tests - 1 1`] = `
"┌ /root
├── a_dir/
├─── a_file.txt
├─── b_file.txt
├─── c_file.txt
"
`;

exports[`JSONToASCII tests - 2 1`] = `
"┌ /root
├── dir1/
├─── dir2/
├──── file3.txt
├─── file1.txt
├─── file2.txt
"
`;

exports[`treeToJSON 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "a_file.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "a_dir",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "c_file.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "b_dir",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "b_file.txt",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "a_dir",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;

exports[`treeToJSON 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "a_file.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "c_file.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "b_file.txt",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "a_dir",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;

exports[`treeToJSON 3`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "file1.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "file2.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file3.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "dir2",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "dir1",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;

exports[`treeToJSON tests - 0 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "a_file.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "a_dir",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "c_file.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "b_dir",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "b_file.txt",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "a_dir",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;

exports[`treeToJSON tests - 1 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "a_file.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "c_file.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "b_file.txt",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "a_dir",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;

exports[`treeToJSON tests - 2 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "file1.txt",
        },
        {
          "children": [],
          "itemType": "file",
          "level": 3,
          "name": "file2.txt",
        },
        {
          "children": [
            {
              "children": [],
              "itemType": "file",
              "level": 4,
              "name": "file3.txt",
            },
          ],
          "itemType": "directory",
          "level": 3,
          "name": "dir2",
        },
      ],
      "itemType": "directory",
      "level": 2,
      "name": "dir1",
    },
  ],
  "itemType": "directory",
  "level": 1,
  "name": "/root",
}
`;
